### PR TITLE
Avoid size overflow when allocating memory

### PR DESCRIPTION
**Pull request by dnakamura**
**Commit 1** - Guard against very large allocation sizes which overflow
when memory tagging overhead is added.

**Commit 2** - omrmemtest: check that large allocations don't overflow.

**Commit 3** - Handle reallocation failures: Restore allocation headers in
the case realloc fails.

Signed-off-by: Devin Nakamura <devinn@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
jenkins build all

**Comment by babsingh** [comment]
> https://ci.eclipse.org/omr/job/PullRequest-linux_ppc-64_le_gcc/3848/consoleFull

Know failure
```
13:06:38  30: [0;32m[----------] [m14 tests from Special/PPCDirectEncodingTest
13:06:38  30: free(): invalid next size (normal)
13:06:38  30/30 Test #30: compunittest ......................Child aborted***Exception:   0.75 sec

**Comment by babsingh** [comment]
Why does the assertion occur? Does it prevent `mem_test8` to complete? 
```
12:59:24  14: [----------] 11 tests from PortMemTest
12:59:24  14: ** ASSERTION FAILED ** omrport.359 at /home/jenkins/jenkins-agent/workspace/Build/port/common/omrmemtag.c:145 Trc_Assert_PRT_memory_corruption_detected((memoryCorruptionDetected))
12:59:24  14: [----------] 11 tests from PortMemTest (26 ms total)
```

**Comment by dnakamura** [comment]
> A new assertion is noticed, which does not appear in previous PR builds. Why does the assertion occur? Does it prevent `mem_test8` to complete?

It does not stop the test from completing. Looking into it, it seems that we clear the allocaiton headers before calling the c-lib realloc, however we never consider the possibility that realloc can fail thus we end up with a still valid chunk of memory, but invalid allocation headers. Note: this isn't specific to this change, but would seem to come into play any time realloc fails for any reason.

**Comment by dnakamura** [comment]
jenkins build all

File: port/common/omrmemtag.c [hunk 1]
```diff
@@ -285,18 +289,26 @@ omrmem_reallocate_memory(struct OMRPortLibrary *portLibrary, void *memoryPointer
...
 			pointer = wrapBlockAndSetTags(portLibrary, pointer, byteAmount, callSite, category);
-		}
-		if (NULL == pointer) {
+		} else {
 			Trc_PRT_mem_omrmem_reallocate_memory_failed_2(callSite, memoryPointer, allocationByteAmount);
+
+			/* Original pointer is still valid, restore mem tags. */
+			if (memoryPointer) {
+				wrapBlockAndSetTags(portLibrary, memoryPointer, savedMemTag.allocSize, savedMemTag.callSite, savedMemTag.category->categoryCode);
+			}
```
Comment (babsingh) [inline port/common/omrmemtag.c:311]:
```suggestion
			/* Original pointer is still valid, restore mem tags. */
			if (NULL != memoryPointer) {
				J9MemTag *savedMemTag = (J9MemTag *)memoryPointer;
				wrapBlockAndSetTags(
						portLibrary,
						memoryPointer,
						savedMemTag->allocSize,
						savedMemTag->callSite,
						savedMemTag->category->categoryCode);
			}
```

File: port/common/omrmemtag.c [hunk 2]
```diff
@@ -285,18 +289,26 @@ omrmem_reallocate_memory(struct OMRPortLibrary *portLibrary, void *memoryPointer
 		omrmem_free_memory(portLibrary, memoryPointer);
 	} else {
 		memoryPointer = unwrapBlockAndCheckTags(portLibrary, memoryPointer);
+		savedMemTag = *((J9MemTag*)memoryPointer);
 		if (NULL == callSite) {
 			/* Inherit the callsite from the original allocation */
-			callSite = ((J9MemTag *) memoryPointer)->callSite;
+			callSite = savedMemTag.callSite;
```
Comment (babsingh) [inline port/common/omrmemtag.c:295]: Restored to the original state.
```suggestion
			callSite = ((J9MemTag *) memoryPointer)->callSite;
```

File: port/common/omrmemtag.c [hunk 3]
```diff
@@ -285,18 +289,26 @@ omrmem_reallocate_memory(struct OMRPortLibrary *portLibrary, void *memoryPointer
 		omrmem_free_memory(portLibrary, memoryPointer);
 	} else {
 		memoryPointer = unwrapBlockAndCheckTags(portLibrary, memoryPointer);
+		savedMemTag = *((J9MemTag*)memoryPointer);
```
Comment (babsingh) [inline port/common/omrmemtag.c:292]:
```suggestion
```

File: port/common/omrmemtag.c [hunk 4]
```diff
@@ -275,6 +278,7 @@ omrmem_reallocate_memory(struct OMRPortLibrary *portLibrary, void *memoryPointer
 {
 	void *pointer = NULL;
 	uintptr_t allocationByteAmount;
+	J9MemTag savedMemTag;
```
Comment (babsingh) [inline port/common/omrmemtag.c:281]:
```suggestion
```

**Comment by babsingh** [comment]
jenkins build all

**Review by babsingh — APPROVED** [review]
> https://ci.eclipse.org/omr/job/PullRequest-linux_ppc-64_le_gcc/3851/

Known failure: https://github.com/eclipse/omr/issues/6571.

> https://github.com/eclipse/omr/pull/6771/checks?check_run_id=8990482324

Known failure: https://github.com/eclipse/omr/issues/6516.

@dnakamura Can you please port these changes to https://github.com/eclipse-openj9/openj9-omr/tree/v0.35.0-release?

**Pull request merged**